coderd/provisionerdserver/provisionerdserver.go

```diff
@@ -543,6 +543,30 @@ func (s *server) acquireProtoJob(ctx context.Context, job database.ProvisionerJo
 			return nil, failJob(fmt.Sprintf("convert workspace transition: %s", err))
 		}
 
+		// A previous workspace build exists
+		var lastWorkspaceBuildParameters []database.WorkspaceBuildParameter
+		if workspaceBuild.BuildNumber > 1 {
+			// TODO: Should we fetch the last build that succeeded? This fetches the
+			//   previous build regardless of the status of the build.
+			buildNum := workspaceBuild.BuildNumber - 1
+			previous, err := s.Database.GetWorkspaceBuildByWorkspaceIDAndBuildNumber(ctx, database.GetWorkspaceBuildByWorkspaceIDAndBuildNumberParams{
+				WorkspaceID: workspaceBuild.WorkspaceID,
+				BuildNumber: buildNum,
+			})
```

Comment:
What problem are we _actually_ trying to solve here? 

`wsbuilder` already fetches the last build parameters, if they exist: https://github.com/coder/coder/blob/af2941bb92306f00c2e7576ec9a0a36a298e603f/coderd/wsbuilder/wsbuilder.go#L650-L669

Given that this is the case, why do we need to do this extra work for _all_ jobs? Isn't this just for template version import jobs?

Reply:
That does it in `coder/coder` at workspace create yes, but this passes the previous values to the terraform via env vars.

The terraform provider now enforces monotonicity: https://github.com/coder/terraform-provider-coder/pull/392

So this is duplicating that check in wsbuilder at `terraform apply/plan`.
For dynamic parameters, we skip validating params in `wsbuilder`, so we need to make sure validation is applied in terraform

Reply:
I like that approach -- `wsbuilder` is then still responsible for fetching all of the various baggage related to a workspace build, but just defers the validation part to Terraform.

Reply:
@johnstcn actually, as much as I'd like to have `wsbuilder` be the source of truth. We fetch everything again at this step in the workspace build.

Current params, workspace data, external auth, etc.

We store very little in the job payload: https://github.com/coder/coder/blob/a7ab88cc4cde1c481332bc0e830f4ff23a5172b4/coderd/provisionerdserver/provisionerdserver.go#L2498-L2504

So I'm going to keep this as a refetching. Ideally I would use the same function to fetch the previous params in both cases, however at `wsbuild` the `latestbuild` is the "previous". And at the point I added code, the previous is `build -1`.

Reply:
It is, I think it would be a large refactor to move all the fields into wsbuilder